File: backend/app/core/workflow_contract.py
from functools import lru_cache
import json
from pathlib import Path
from typing import Any

from pydantic import BaseModel, Field

from app.core.config import get_settings


class MappingRule(BaseModel):
    name: str
    status: str
    when: dict[str, Any]


class PollingProgressSchema(BaseModel):
    required: list[str]
    properties: dict[str, str]


class WorkflowContract(BaseModel):
    version: str
    visible_statuses: list[dict[str, str]]
    internal_states: list[str]
    failure_reasons: list[str]
    workflow_kinds: list[str]
    mapping_rules: list[MappingRule]
    polling_progress_schema: PollingProgressSchema = Field(alias="polling_progress_schema")


@lru_cache
def get_workflow_contract() -> WorkflowContract:
    settings = get_settings()
    configured_path = Path(settings.shared_contract_path)
    fallback_paths = (
        Path(__file__).resolve().parents[2] / "shared" / "workflow-contract.json",
        Path(__file__).resolve().parents[3] / "shared" / "workflow-contract.json",
    )
    candidate_paths = (configured_path, *fallback_paths)

    contract_path = next((path for path in candidate_paths if path.exists()), None)
    if contract_path is None:
        raise FileNotFoundError(
            "Workflow contract file not found. Checked: "
            + ", ".join(str(path) for path in candidate_paths)
        )

    return WorkflowContract.model_validate(json.loads(contract_path.read_text()))

File: frontend/src/lib/workflow-contract.ts
import { z } from "zod";
import contract from "./workflow-contract.json";

const visibleStatusSchema = z.object({
  id: z.string(),
  label: z.string(),
});

const workflowContractSchema = z.object({
  version: z.string(),
  visible_statuses: z.array(visibleStatusSchema),
  internal_states: z.array(z.string()),
  failure_reasons: z.array(z.string()),
  workflow_kinds: z.array(z.string()),
  mapping_rules: z.array(
    z.object({
      name: z.string(),
      status: z.string(),
      when: z.record(z.unknown()),
    }),
  ),
  polling_progress_schema: z.object({
    required: z.array(z.string()),
    properties: z.record(z.string()),
  }),
});

export const workflowContract = workflowContractSchema.parse(contract);

export type VisibleStatusId = (typeof workflowContract.visible_statuses)[number]["id"];

export const visibleStatuses = workflowContract.visible_statuses;
